lib/AST/LocalizationFormat.cpp

```diff
@@ -143,31 +137,50 @@ void YAMLLocalizationProducer::forEachAvailable(
   }
 }
 
+bool readID(llvm::yaml::IO &io, uint32_t &id,
+            std::vector<std::string> &unknownIDs) {
+  LocalDiagID diagID;
+  io.mapRequired("id", diagID);
+  if (diagID == LocalDiagID::NumDiags) {
+    std::string idName;
+    io.mapRequired("id", idName);
+    unknownIDs.push_back(idName);
+    return true;
+  } else {
+    id = static_cast<uint32_t>(diagID);
+    return false;
+  }
+}
+
 template <typename T, typename Context>
 typename std::enable_if<llvm::yaml::has_SequenceTraits<T>::value, void>::type
-readYAML(llvm::yaml::IO &io, T &Seq, bool, Context &Ctx) {
+readYAML(llvm::yaml::IO &io, T &Seq, T &unknownIDs, bool, Context &Ctx) {
   unsigned count = io.beginSequence();
-  if (count)
+  if (count) {
     Seq.resize(LocalDiagID::NumDiags);
+  }
+
+  DiagnosticNode current;
   for (unsigned i = 0; i < count; ++i) {
     void *SaveInfo;
     if (io.preflightElement(i, SaveInfo)) {
-      DiagnosticNode current;
-      yamlize(io, current, true, Ctx);
-      io.postflightElement(SaveInfo);
+      io.beginMapping();
+
+      // If the current diagnostic ID is available in YAML but not available in
+      // `.def`, add it to the `unknownIDs` array and continue. Otherwise, parse
+      // the current diagnnostic message and store it in the diagnostics array.
+      if (!readID(io, current.id, unknownIDs)) {
```

Comment:
@xedin I think that what was in your mind, am I close?

Reply:
Isn't `msg` has to be read regardless whether `id` is found or not? The format is `id` and `msg`, can parser just skip to the next block without reading the rest of the current one?

Reply:
Yes, it can skip the block. I found out that it's not mandatory to read the message.

Reply:
I also ran the tests and everything seemed fine.

Reply:
Interesting, that's pretty lucky although I wonder how it's possible to have mismatched `beginMapping` and `endMapping`, it seems weird that it's possible to begin mapping again without closing it first...

Reply:
Actually, I didn't mismatch `beginMapping` and `endMapping`. If I ended the mapping after L182 https://github.com/apple/swift/pull/33337/files/a4776146da56b733557ec710bdb6b3e64121841d#diff-594606b3ff2e7321e3fac5cc36c08dcbR182 I will get an error that I didn't read `msg` to be precise I'll get this error `error: unknown key 'msg'`. 

Here's what I realized (might be wrong)... what happens is, I start mapping `id` and if the `id` isn't available, the for loop will continue to next block and will start a new mapping, etc.

Reply:
What I mean is - `endMapping` is only called for a valid id, otherwise we’d up calling `beginMapping` without closing it first. I don’t really get it why not just read `msg` unconditionally here.